text/0000-custom-panic-handlers.md

```diff
@@ -0,0 +1,80 @@
+
+`std::rt::unwind::begin_unwind_inner` would call the thread's panic handler before processing callbacks, without needing synchronization.
+
+If the user wishes not to log panics, he can define an empty function following the `Callback` signature and pass it to `set_panic_handler`.
+
+### Summary of the proposed changes
+
+* Add a thread-local handler to `rt::unwind`, defaulting to the [current default panic handler](https://github.com/rust-lang/rust/blob/2b8c9b12f91c0bf2c1e6278a5f803c2df3698432/src/libstd/panicking.rs#L28):
+```rust
+thread_local! { static ON_PANIC: Cell<Callback> = Cell::new(panicking::on_panic) }
+```
+
+* Add a function to `std::rt::unwind` or `std::thread` (see unresolved questions) to set a new panic handler:
+``` rust
+pub fn set_panic_handler(new_handler: Callback) {
+  ON_PANIC.with(|cb_cell| cb_cell.set(new_handler));
+}
+```
+* Replace the `unsafe` block registering the default callback in [`std::rt::unwind::begin_unwind_inner`](https://github.com/rust-lang/rust/blob/9cc0b2247509d61d6a246a5c5ad67f84b9a2d8b6/src/libstd/rt/unwind/mod.rs#L241-L254) with a plain call to the thread's handler:
+``` rust
+let (file, line) = *file_line;
+ON_PANIC.with(|cb_cell| cb_cell.get()(&*msg, file, line));
+```
+
+
+# Drawbacks
+
+This proposed solution does not implement inheritance of panic handlers between threads (that is, if thread A has the handler `foo` and spawns a second thread B, B will also have `foo` as a handler).
+However, this can be implemented later in `std::thread` and it should be relatively easy to do so (read the calling thread's panic handler and set the new thread's handler to it). For now, it is probably best to keep the changes minimal.
+
+
+# Alternatives
+
+The original solution I proposed was to just make `panic!()` output nothing to `stderr` when called without arguments, instead of calling `panic!("explicit panic")` internally. This is simpler to implement and requires less changes to `rt::unwind`. However, it turned out a more flexible implementation would be better to support most use cases; one does not always have control over the functions he calls, and hence cannot prevent them from panicking and logging a debug message without rewriting their code.
+
+We could also not change anything to the panicking mechanisms, use `catch_panic` to react differently on panics and ignore unwanted panic messages, or run the panicking process from a wrapper which removes all "thread {} panicked" entries from the output of the program before forwarding it. However, this approach is much more cumbersome than the proposed solution and much less efficient as it requires running a new wrapper process.
+
+In case one simply wishes to prevent logging of panics in a specific thread entirely, they can also use [this workaround](https://github.com/rust-lang/rust/issues/24099#issuecomment-89908401).
+
+
+# Unresolved questions
+
+Should `set_panic_handler` live in `std::rt`, or in `std::thread`?
```

Comment:
both this and `std::thread::catch_panic` should maybe move out of `std::thread` though